packages/zoe/src/contractSupport/zoeHelpers.js

```diff
@@ -3,438 +3,363 @@
-     */
-    trade: (keepLeft, tryRight) => {
-      assert(
-        keepLeft.offerHandle !== tryRight.offerHandle,
-        details`an offer cannot trade with itself`,
-      );
-      let leftAllocation = zcf.getCurrentAllocation(keepLeft.offerHandle);
-      let rightAllocation = zcf.getCurrentAllocation(tryRight.offerHandle);
-
-      try {
-        // for all the keywords and amounts in leftGains, transfer from
-        // right to left
-        ({ from: rightAllocation, to: leftAllocation } = calcNewAllocations(
-          { from: rightAllocation, to: leftAllocation },
-          keepLeft.gains,
-          tryRight.losses,
-        ));
-        // For all the keywords and amounts in rightGains, transfer from
-        // left to right
-        ({ from: leftAllocation, to: rightAllocation } = calcNewAllocations(
-          { from: leftAllocation, to: rightAllocation },
-          tryRight.gains,
-          keepLeft.losses,
-        ));
-      } catch (err) {
-        return rejectOffer(tryRight.offerHandle);
-      }
+/** @type {Trade} */
+export const trade = (zcf, keepLeft, tryRight) => {
+  assert(
+    keepLeft.seat !== tryRight.seat,
+    details`an offer cannot trade with itself`,
+  );
+  let leftAllocation = keepLeft.seat.getCurrentAllocation();
+  let rightAllocation = tryRight.seat.getCurrentAllocation();
+  try {
+    // for all the keywords and amounts in leftGains, transfer from
+    // right to left
+    ({ from: rightAllocation, to: leftAllocation } = calcNewAllocations(
+      zcf,
+      { from: rightAllocation, to: leftAllocation },
+      keepLeft.gains,
+      tryRight.losses,
+    ));
+    // For all the keywords and amounts in rightGains, transfer from
+    // left to right
+    ({ from: leftAllocation, to: rightAllocation } = calcNewAllocations(
+      zcf,
+      { from: leftAllocation, to: rightAllocation },
+      tryRight.gains,
+      keepLeft.losses,
+    ));
+  } catch (err) {
+    console.log(err);
+    throw tryRight.seat.kickOut();
+  }
+
+  // Check whether reallocate would error before calling. If
+  // it would error, reject the right offer and return.
+  const offerSafeForLeft = keepLeft.seat.isOfferSafe(leftAllocation);
+  const offerSafeForRight = tryRight.seat.isOfferSafe(rightAllocation);
+  if (!(offerSafeForLeft && offerSafeForRight)) {
+    console.log(`currentLeftAllocation`, keepLeft.seat.getCurrentAllocation());
+    console.log(`currentRightAllocation`, tryRight.seat.getCurrentAllocation());
+    console.log(`proposed left reallocation`, leftAllocation);
+    console.log(`proposed right reallocation`, rightAllocation);
+    // show the constraints
+    console.log(`left want`, keepLeft.seat.getProposal().want);
+    console.log(`right want`, tryRight.seat.getProposal().want);
+
+    if (!offerSafeForLeft) {
+      console.log(`offer not safe for left`);
+    }
+    if (!offerSafeForRight) {
+      console.log(`offer not safe for right`);
+    }
+    return tryRight.seat.kickOut();
+  }
```

Comment:
The `kickOut` call should have some informative error message. Given the wealth of diagnostics we're sending to the log for this case, the message should suggest checking the log.

Regarding the logging, what is our current practice re `console.log` calls in production code? I think it is a good idea, but I haven't see much elsewhere. For zoe, all of then are in this vicinity. Since these are only about a detected mistake, should they be `console.warn`? We should revisit such questions during our upcoming logging summit.